agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -62,17 +67,71 @@ def on_container_event(topic, e)
       if attrs['io.kontena.service.id'] == @service_pod.service_id &&
           attrs['io.kontena.service.instance_number'].to_i == @service_pod.instance_number
         @container_state_changed = true
+        handle_restart_on_die if e.status == 'die'.freeze
+      end
+    end
+
+    # Handles events when container has died
+    def handle_restart_on_die
+      unless @service_pod.running?
+        @restart_backoff_timer.cancel if @restart_backoff_timer
+        return
       end
+
+      return if apply_in_progress?
+
+      # cancel counters because we are called again
+      cancel_restart_timers
+
+      # backoff restarts
+      backoff = @restarts ** 2
+      backoff = max_restart_backoff if backoff > max_restart_backoff
+      if backoff == 0
+        info "restarting #{@service_pod.name_for_humans} because it has stopped"
+      else
+        info "restarting #{@service_pod.name_for_humans} because it has stopped (delay: #{backoff}s)"
+      end
+      @restart_backoff_timer = after(backoff) {
+        apply_restart if @container_state_changed
+      }
+      @restarts += 1
+    end
+
+    def max_restart_backoff
+      Kontena::Workers::ServicePodManager::LOOP_INTERVAL
+    end
+
+    def cancel_restart_timers
+      @restart_counter_reset_timer.cancel if @restart_counter_reset_timer
+      @restart_backoff_timer.cancel if @restart_backoff_timer
+    end
+
+    def apply_restart
+      apply
+      # reset restart counter if instance stays up 10s
+      @restart_counter_reset_timer = after(10) {
+        debug "#{@service_pod.name_for_humans} stayed up 10s, resetting restart backoff counter"
+        @restarts = 0
+      }
+    end
+
+    def restarting?
+      @restarts > 0
     end
 
     def destroy
       @service_pod.mark_as_terminated
       apply
     end
 
+    def apply_in_progress?
```

Comment:
Remove the `@apply_in_progress` if the apply is exclusive, it doesn't work and it's misleading.